### PR TITLE
Fixed Extend RangeInterface to allow getRangeType() calls on range object

### PR DESCRIPTION
**Pull request by Empornium**

### CODE REVIEW

**Comment by mlocati** [comment]
I don't think that adding is method is logically correct.
For instance, what should be the result of
```php
$range = \IPLib\Factory::rangeFromString('*.*.*.*');
$range->getRangeType()
```

?

**Comment by Empornium** [comment]
At the moment that would call getRangeType() on 0.0.0.0, which would return T_UNSPECIFIED. Which seems sensible enough.

You have a point that some ranges may cross many types, and perhaps T_UNSPECIFIED or similar should be returned under such circumstances, we'd be willing to submit a PR with such a change.

It seems to us that having the ability to check the type of a range would be a useful addition.

**Comment by mlocati** [comment]
> It seems to us that having the ability to check the type of a range would be a useful addition.

I do agree, provided that *all* the IP in the range are of the same type.

> You have a point that some ranges may cross many types, and perhaps T_UNSPECIFIED or similar should be returned under such circumstances.

What about returning `null` if the range contains mixed types? So that users have a way to know that the range contains mixed types.

> we'd be willing to submit a PR with such a change

You can update this pull request (#12) itself.

**Comment by Empornium** [comment]
> What about returning null if the range contains mixed types? So that users have a way to know that the range contains mixed types.

Seems sensible, we'll go with that.

> You can update this pull request (#12) itself.

Will do. :)

**Comment by Empornium** [comment]
The way we went with this ended up being more invasive than I first imagined. If this is too big a change or if any of the style isn't as you'd like it please just say and we'll of course make any alterations you'd like.

The windfall of the way we've ended up implementing this is that Range::contains() can now check if a range contains an address or another range, which also seems like an appealing addition.

File: src/Address/IPv4.php [hunk 1]
```diff
@@ -41,6 +41,30 @@ class IPv4 implements AddressInterface
      */
     protected $comparableString;
 
+    /**
+     * The list of IPv4 RFC designated address ranges.
+     *
+     * @var mixed
+     */
+    public static $reservedRanges = array(
```
Comment (mlocati) [inline src/Address/IPv4.php]: What about making this private and caching its values?
I mean, something like this:
```php
/*
 * @var array|null
 */
private static $reservedRanges = null;
/*
 * @return array
 */
public static getReservedRanges()
{
    if (self::$reservedRanges === null) {
        $reservedRanges = array();
        foreach(array(
            '0.0.0.0/32' => RangeType::T_UNSPECIFIED),      //RFC 5735
            '0.0.0.0/8' => RangeType::T_THISNETWORK),      //RFC 5735
            // ...
        ) as $range => $type) {
            $reservedRanges[] = array('range' => Subnet::fromString($range), 'type' => $type);
        }
        self::$reservedRanges = $reservedRanges;
    }
    return self::$reservedRanges;
}
```

File: src/Address/IPv6.php [hunk 2]
```diff
@@ -50,6 +50,38 @@ class IPv6 implements AddressInterface
      */
     protected $rangeType;
 
+    /**
+     * The list of IPv6 RFC designated address ranges.
+     *
+     * @var mixed
+     */
+    public static $reservedRanges = array(
```
Comment (mlocati) [inline src/Address/IPv6.php]: Same comment as `IPv4::$reservedRanges`

File: src/Address/IPv6.php [hunk 3]
```diff
@@ -50,6 +50,38 @@ class IPv6 implements AddressInterface
...
+        array('cidr' => '::/128',    'type' => RangeType::T_UNSPECIFIED),        //RFC 4291
+        array('cidr' => '::1/128',   'type' => RangeType::T_LOOPBACK),           //RFC 4291
+        array('cidr' => '100::/64',  'type' => RangeType::T_DISCARDONLY),        //RFC 4291
+        array('cidr' => '100::/8',   'type' => RangeType::T_DISCARD),            //RFC 4291
+        //['cidr' => '2002::/16', 'type' => RangeType::],                   //RFC 4291
```
Comment (mlocati) [inline src/Address/IPv6.php]: What are these commented lines?

File: src/Range/Pattern.php [hunk 4]
```diff
@@ -143,20 +153,94 @@ public function getAddressType()
...
+
+        return $this->rangeType;
+    }
+
     /**
      * {@inheritdoc}
      *
      * @see RangeInterface::contains()
      */
-    public function contains(AddressInterface $address)
```
Comment (mlocati) [inline src/Range/Pattern.php]: I'd add a new method to work with ranges, for these reasons:
1. if someone extends this class, we have a backward compatibility break (the method signature changes) - see for example what would happen: https://3v4l.org/Edubs
2. keep things as simple as possible: one method to do one thing

File: src/Range/RangeInterface.php [hunk 5]
```diff
@@ -34,14 +34,21 @@ public function __toString();
...
+
     /**
      * Check if this range contains an IP address.
      *
      * @param AddressInterface $address
      *
      * @return bool
      */
-    public function contains(AddressInterface $address);
+    public function contains($address);
```
Comment (mlocati) [inline src/Range/RangeInterface.php]: See comment for Pattern::contains()

File: src/Range/Single.php [hunk 6]
```diff
@@ -88,16 +88,33 @@ public function getAddressType()
...
+        return $this->address->getRangeType();
+    }
+
     /**
      * {@inheritdoc}
      *
      * @see RangeInterface::contains()
      */
-    public function contains(AddressInterface $address)
+    public function contains($address)
```
Comment (mlocati) [inline src/Range/Single.php]: See comment for Pattern::contains()

File: src/Range/Subnet.php [hunk 7]
```diff
@@ -125,20 +136,94 @@ public function getAddressType()
...
+        return $this->rangeType;
+    }
+
     /**
      * {@inheritdoc}
      *
      * @see RangeInterface::contains()
      */
-    public function contains(AddressInterface $address)
+    public function contains($address)
```
Comment (mlocati) [inline src/Range/Subnet.php]: See comment for Pattern::contains()

File: test/tests/Addresses/AddressTypeTest.php [hunk 8]
```diff
@@ -6,7 +6,7 @@
 use IPLib\Address\Type;
 use PHPUnit_Framework_TestCase;
 
-class AddressTypeTest extends PHPUnit_Framework_TestCase
+class RangeRangeTypeTest extends PHPUnit_Framework_TestCase
```
Comment (mlocati) [inline test/tests/Addresses/AddressTypeTest.php]: Why did you change the class name?
Furthermore, let's follow the PSR-4 rules: the class name must be the same as the file name (except for the leading .php file extension).

File: test/tests/Addresses/RangeTypeTest.php [hunk 9]
```diff
@@ -6,7 +6,7 @@
 use IPLib\Range\Type;
 use PHPUnit_Framework_TestCase;
 
-class TypeTest extends PHPUnit_Framework_TestCase
+class AddressRangeTypeTest extends PHPUnit_Framework_TestCase
```
Comment (mlocati) [inline test/tests/Addresses/RangeTypeTest.php]: Why did you change the class name?
Furthermore, let's follow the PSR-4 rules: the class name must be the same as the file name (except for the leading .php file extension).

**Comment by mlocati** [comment]
The coverage dropped from 99.03 to 91.53.
Maybe it's caused by coveralls not correctly covering arrays, but it's worth investigating that IMHO.

File: src/Address/IPv4.php (hunk 1 above)
Comment (Empornium) [reply src/Address/IPv4.php]: Seems sensible to me. :)

File: src/Address/IPv6.php (hunk 3 above)
Comment (Empornium) [reply src/Address/IPv6.php]: Commented out for clarity (maybe not clear), they are reserved ranges as defined in the RFC but as the default is T_RESERVED for IPv6 there's no need to perform the checks in the loop. So commenting them out is a performance improvement, if you like we could uncomment them for completeness.

File: src/Range/Pattern.php (hunk 4 above)
Comment (Empornium) [reply src/Range/Pattern.php]: I agree, a separate function is cleaner.

File: test/tests/Addresses/AddressTypeTest.php (hunk 8 above)
Comment (Empornium) [reply test/tests/Addresses/AddressTypeTest.php]: Class name collision caused composer to complain and fail. Tests aren't complete yet as Pattern testing isn't covered, we'll of course lean up all the file names etc when we're at that stage so consider that very much WIP.

**Comment by Empornium** [comment]
Yeah, there's fairly bad coverage on Pattern.php now as none of the tests target it. Now we know what direction you'd like us to go we'll work on finishing this off for next week.

File: test/tests/Addresses/AddressTypeTest.php (hunk 8 above)
Comment (mlocati) [reply test/tests/Addresses/AddressTypeTest.php]: > Class name collision caused composer to complain and fail

Since we use namespaces (that reflects the folder structure), collisions occur only in case the fully qualified class names (that is, namespace+class, like IPLib\Test\Addresses\RangeTypeTest) are the same.
But since we should follow PSR-4 (the namespace is tied to the folder and the class name is tied to the file name), we shouldn't have collisions... What are the fully qualified names of the classes that collides?

Comment (Empornium) [reply test/tests/Addresses/AddressTypeTest.php]: https://travis-ci.org/mlocati/ip-lib/jobs/315451142

`Fatal error: Cannot declare class IPLib\Test\Addresses\TypeTest, because the name is already in use in /home/travis/build/mlocati/ip-lib/test/tests/Ranges/RangeTypeTest.php on line 296`

However, I just noticed the Range TypeTest is in the wrong namespace... I blame lack of coffee on that one. I'll get it fixed up.

:)

Comment (mlocati) [reply test/tests/Addresses/AddressTypeTest.php]: 😆 👍

**Comment by Empornium** [comment]
I think this is as high as I can get the coverage tests to go, we're also not sure what to do with the issue StyleCI is reporting.

**Comment by mlocati** [comment]
Thank you very much!

File: src/Range/Pattern.php [hunk 10]
```diff
@@ -143,20 +153,95 @@ public function getAddressType()
...
+
+            // Check if range is contained within an RFC subnet
+            foreach ($reservedRanges as $reservedRange) {
+                if ($reservedRange['range']->containsRange($this)) {
+                    $this->rangeType = $reservedRange['type'];
+                    break;
+                }
+            }
+
+            // Check if public/reserved range contains an RFC subnet
```
Comment (mlocati) [inline src/Range/Pattern.php]: @Empornium could you explain a bit the logic if this piece of code (from line 191 to line 214)?

Comment (Empornium) [reply src/Range/Pattern.php]: The default for IPv4 is "Public" and for IPv6 is "Reserved", at this point in the function we've already checked if the supplied pattern is contained in any defined ranges so it should have "Public" or "Reserved" if it is not contained within a range returned by getReservedRanges().

After determining it isn't contained within one of the defined reserved ranges we need to determine if it contains any reserved ranges (with a few exceptions) and return null if it does.

For instance, 192.166.0.0/8 is not contained within any defined ranges and as such will have a RangeType of T_PUBLIC at this point in the function, it does however contain within in 192.168.0.0/16 and so should return null at the end.

The specific exceptions are defined in the various RFCs.

Does that make sense?

Comment (mlocati) [reply src/Range/Pattern.php]: Thanks for the explanation. I already simplified this approach (see the current code in the master branch), so that it's easier to underatand and to maintain. And I also fixed some errors in your approach (see the changes I did on your tests).

Comment (Empornium) [reply src/Range/Pattern.php]: Classy.
